src/MCPServer/lib/server/queues.py

```diff
@@ -221,34 +197,33 @@ def stop(self):
         self.shutdown_event.set()
 
     def _package_completed_callback(self, package, future):
-        """Called by an Executor on completion the last job in a package.
+        """Mark the package as inactive and schedule a new package.
 
-        The result of the future will always be None.
+        It is assumed that a package is only complete when a terminal link is
+        hit. When we hit a terminal link the future should not contain the next
+        job in the chain. If the future is a new job we raise an exception.
         """
         if future.result() is not None:
             raise RuntimeError(
                 "Unexpectedly received another job on package completion. "
-                "If the workflow has been modified, please update "
-                "PACKAGE_COMPLETED_LINK_IDS."
+                "Please verify the value of `end` in the workflow. "
             )
 
         self.deactivate_package(package)
         self.queue_next_job()
 
     def _job_completed_callback(self, future):
-        """Called by an Executor on completion of `job.run`.
+        """Schedule the next job in the chain.
 
-        The argument is a future, the result of which should be the next job
-        in the chain.
+        Retrieves the next job from the result from the previous job.
+        If there is no next_job return, otherwhise schedule the new job.
         """
         metrics.active_jobs_gauge.dec()
         next_job = future.result()
 
         if not next_job:
             return
-
-        # Special case for decision job here.
-        if isinstance(next_job, DecisionJob) and next_job.awaiting_decision:
+        elif isinstance(next_job, DecisionJob) and next_job.awaiting_decision:
```

Comment:
I see what you mean. The reason I changed it was because to me it all belongs to the same context, I see it as a switch on `next_job`. If there is no `next_job` we do nothing. In case we have a decision job we await the decision. In all other cases, we schedule the next job. 

The separation made me wonder why the early return was a special case. To me, it makes more sense now but if you prefer the old way I am more than happy to change it back,